File: src/cxx-dev/LiENa/liena.cpp
#include "liena.h"


/**
 *
 *  @brief liena::liena
 *
 */
liena::liena(){
    this->init();
    this->connexion();
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::~liena
//!
liena::~liena(){
    delete this->globalParameter;
    delete this->tcpServer;
    delete this->outputMessageCache;
    delete this->inputMessageCache;
    delete this->distributedSystem;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtCommunicationStack::init
//!
void liena::init(){

    this->globalParameter    = new lienaGlobal();                                    // pre-defined essential parameter
    this->tcpServer          = new lienaTcpServer(this->globalParameter);            // tcp server pour ecouter des autres composants du system robotic
    this->inputMessageCache  = new lienaInputMessageCache();                         // input message cache for developer
    this->outputMessageCache = new lienaOutputMessageCache();                        // output message cache for developer
    this->distributedSystem  = new lienaDistributedSystem(this->globalParameter);    // object to store all module in the distributed system
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::setConnections
//!
void liena::connexion(){
    qRegisterMetaType<SOCKET>("SOCKET");
    this->connect(this->tcpServer,         SIGNAL(clientArrived(SOCKET, QString)),       this, SLOT(createReceptionChannel(SOCKET, QString)));
    this->connect(this->tcpServer,         SIGNAL(localIPDetect(QString)),               this, SLOT(getSelfIp(QString)));
    this->connect(this->distributedSystem, SIGNAL(generateNewMessageSequence(uint32_t)), this, SLOT(generateNewMessageSequenceBy(uint32_t)));
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::generateNewMessageSequenceBy
//! \param deviceID
//!
void liena::generateNewMessageSequenceBy(uint32_t deviceID){
    qDebug()<<"liena::generateNewMessageSequenceBy"<<deviceID;
    this->distributedSystem->configureMessageQueuePairBy(deviceID, this->outputMessageCache->generateNewMessageSequence(deviceID), this->inputMessageCache->generateNewMessageSequence(deviceID));
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::registDevice
//!
void liena::registerDevices(){
   this->globalParameter->setLocalDeviceId(this->globalParameter->getDSI()->get_first_deviceId());
   this->distributedSystem->setRoboticSystemModule(this->globalParameter->getDSI()->get_deviceIds());
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::setLocalDeviceID
//! \param deviceID
//!
void liena::setLocalDeviceID(uint32_t deviceID){
    this->globalParameter->setLocalDeviceId(deviceID);
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::checkAvailability
//! \param deviceID
//! \return
//!
bool liena::checkAvailability(uint32_t deviceID){
    bool ret = false;
    return ret;
    //this->inputMessageCache->write
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtCommunicationStack::lauchServer
//! \return
//!
bool liena::launch(){
    return this->tcpServer->launchServer();
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtCommunicationStack::stopServer
//! \return
//!
bool liena::terminate(){
    bool ret = false;
    ret = this->tcpServer->stopServer();
    return ret;
}

//!--------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::incomingConnection
//! \param handle : qintptr
//!
void liena::createReceptionChannel(SOCKET socketReception, QString addr){
    if(this->distributedSystem->checkModuleByAddr(addr)){
        this->distributedSystem->generateReceptionChannelByAddr(socketReception, addr);
    }
    else{
        this->distributedSystem->createDistributedModuleForReceptionChannel(socketReception);
    }
}

//!--------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief createDistributedModuleForTransmissionChannel
//! \param addr
//! \param port
//!
void liena::createMotivateModule(uint32_t targetDeviceId, QString addr,uint16_t port){
    this->distributedSystem->createDistributedModuleWithTransmissionChannel(targetDeviceId, true, addr, port);
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::openChannel : programming interface for developer, aim to open a in&out channel which a target device/software
//! \param addr
//! \param port
//! \return
//!
bool liena::openSessionRequest(uint32_t targetDeviceId, QString addr,uint16_t port){
    if(!this->tcpServer->getStatus()){
        qDebug()<<"liena tcp server not launched ...";
        return false;
    }

    this->createMotivateModule(targetDeviceId, addr, port);

    return true;
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::closeSessionRequest
//! \param deviceID
//!
void liena::closeSessionRequest(uint32_t deviceID){
    this->distributedSystem->closeDistributedChannel(deviceID);
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::getInputMessageCache
//! \return
//!
lienaInputMessageCache* liena::getInputMessageCache(){
    return this->inputMessageCache;
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief liena::getOutputMessageCache
//! \return
//!
lienaOutputMessageCache* liena::getOutputMessageCache(){
    return this->outputMessageCache;
}

//! --------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtCommunicationStack::getSelfIp
//! \param addr
//!
void liena::getSelfIp(QString addr){
    this->globalParameter->setLocalIPAddr(addr);
}

//! --------------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtCommunicationStack::connectionEstablish
//!
void liena::connectionEstablish(){
    //this->transmissionTaskManager->launchLatestTask();
    //this->addHelloMessageTask->start();
}

File: src/py-dev/LiENa/liena.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
from LiENa.LiENaBasic.lienaGlobal import LienaGlobal
from LiENa.LiENaDistributedSystem.lienaDistributedSystem import LienaDistributedSystem
from LiENa.LiENaStructure.LiENaMessageStructure.lienaInputMessageCache import LienaInputMessageCache
from LiENa.LiENaStructure.LiENaMessageStructure.lienaOutputMessageCache import LienaOutputMessageCache
from LiENa.LiENaSocket.LienaTcpServer import LienaTcpServer
from LiENa.LiENaBasic.lienaDefinition import *
from PyQt5.QtCore import QObject, pyqtSignal
import time


class Liena(QObject):
    """
          description:
    """
    newHandShakeMsgArrived = pyqtSignal()
    newConnection = pyqtSignal()

    # [0] ---------------------------------------------------------------------------------------------------------------------------
    def __init__(self):
        super(Liena, self).__init__()

        self.system_flag = False

        # the class which contain import parameter of the middleware
        self.globalParameter = LienaGlobal()

        # the global tcp server waiting for motivate incoming distributed module(remote networked device)
        self.tcpServer = LienaTcpServer(self.globalParameter)

        # the class for manage all the real time session ( message exchange, scheduling, failure handling)
        self.distributedSystem = LienaDistributedSystem(self.globalParameter)

        #
        self.inputMessageCache = LienaInputMessageCache()
        #
        self.outputMessageCache = LienaOutputMessageCache()

        # signals/slots connexion
        self.distributedSystem.needToRestartServer.connect(self.restart_tcp_server)
        self.tcpServer.clientArrived.connect(self.create_reception_channel_by_address)
        # self.distributedSystem.generateNewMessagesequence.con

        self.distributedSystem.generateNewMessageSequence[int].connect(self.generate_new_msg_sequence_by)

    def generate_new_msg_sequence_by(self, device_id):
        self.distributedSystem.configure_msg_sequence_pair_by(device_id, self.outputMessageCache.generate_new_msg_seq(device_id), self.inputMessageCache.generate_new_msg_seq(device_id))

    # [0] ---------------------------------------------------------------------------------------------------------------------------
    def set_local_device_id(self, local_device_id):
        self.globalParameter.set_local_device_id(local_device_id)

    # [1] ---------------------------------------------------------------------------------------------------------------------------
    def register_device(self, local_device_id, device_indexes):
        """
            -- declare all distributed modules of a medical robotic distributed system by passing their unique device id

        :param local_device_id:  device id of the local module
        :param device_indexes:   all device id set of a medical robotic
        :return:
        """
        self.globalParameter.set_local_device_id(local_device_id)
        self.distributedSystem.set_robotic_system_module(device_indexes)

    # [2] ---------------------------------------------------------------------------------------------------------------------------
    def launch(self):
        """
            -- launch liena global server wait for motivate incoming distributed module
        :return:
        """
        if not self.system_flag:
            self.tcpServer.launch_server()
            self.system_flag = True

    # [3] ---------------------------------------------------------------------------------------------------------------------------
    def open_session_request(self, target_device_id, target_ip_address, port):
        """
            -- start a real time communication session with the device using target_device_id, target ip address
        :param target_device_id: 
        :param target_ip_address: 
        :param port: 
        :return: 
        """

        # verifier if the global server is launched
        if not self.system_flag:
            self.launch()

        # generate the session and the outgoing pipeline of the session
        self.distributedSystem.create_distributed_module_with_transmission_channel(target_device_id, True, target_ip_address, port)

    # [4] ---------------------------------------------------------------------------------------------------------------------------
    def close_session_request(self, device_id):
        """
            -- close a real time communication session with the device indicated
        :param device_id:
        :return:
        """
        self.distributedSystem.close_session_by_device_id(device_id)

    # [5] ---------------------------------------------------------------------------------------------------------------------------
    def create_reception_channel_by_address(self):
        incoming_socket = self.tcpServer.get_latest_socket()

        if self.distributedSystem.check_module_by_addr(incoming_socket[1]):
            self.distributedSystem.generate_reception_channel_by_addr(incoming_socket[1], incoming_socket[0])
        else:
            self.distributedSystem.create_distributed_module_with_reception_channel(incoming_socket[0])


    # [6] ---------------------------------------------------------------------------------------------------------------------------
    def terminate(self):
        if self.system_flag:
            self.tcpServer.terminate_server()
            self.system_flag = False

    # [7] ---------------------------------------------------------------------------------------------------------------------------
    def restart_tcp_server(self):
        if DEBUG:
            print("close server")
        self.tcpServer.close()
        time.sleep(1)
        if DEBUG:
            print("restart server")
        self.tcpServer.restart()

    def get_input_cache(self):
        return self.inputMessageCache

    def get_output_cache(self):
        return self.outputMessageCache

    def close(self):
        self.context.close_system()
        self.tcpServer.terminate_server()

    # def launch_transmission_task_by_addr(self, addr):
    #     for transmissionTask in self.transmissionTaskManager:
    #         if transmissionTask.get_addr() == addr:
    #             transmissionTask.launch()

    # deprecated
    def create_motivate_module(self, device_id, addr, port):
        self.distributedSystem.create_distributed_module_with_transmission_channel(device_id, True, addr, port)

